profileservice/dynamodb.go

```diff
@@ -0,0 +1,282 @@
+package profileservice
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/CollActionteam/collaction_backend/auth"
+	"github.com/aws/aws-lambda-go/events"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+)
+
+var (
+	tablename = os.Getenv("PROFILE_TABLE")
+)
+
+func connDb() (svc *dynamodb.DynamoDB) {
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+
+	svc = dynamodb.New(sess)
+	return
+}
+
+func UpdateProfile(req events.APIGatewayV2HTTPRequest) (err error) {
+	var profiledata = Profile{}
+	var skipFields = []string{"userid", "displayname", "phone"}
+
+	svc := connDb()
+
+	usrInf, err := auth.ExtractUserInfoV2(req)
+	if err != nil {
+		return err
+	}
+
+	userID := usrInf.UserID()
+
+	err = json.Unmarshal([]byte(req.Body), &profiledata)
+	if err != nil {
+		return err
+	}
+
+	err = profiledata.ValidateProfileStruct("update")
+	if err != nil {
+		return err
+	}
+
+	requiredMap := map[string]string{}
+
+	inrec, _ := json.Marshal(profiledata)
+	json.Unmarshal(inrec, &requiredMap)
+
+	nw := len(requiredMap)
+
+	if nw < 1 {
+		return fmt.Errorf("no user data provided")
+	}
+
+	var wg sync.WaitGroup
+
+	wg.Add(nw)
+
+	wrkchan := make(chan error, nw)
+
+	for i, v := range requiredMap {
+		if contains(i, skipFields) || v == "" {
+			wg.Done()
+			continue
+		} else {
+			go UpdateProfileTableItem(i, v, userID, svc, wrkchan, &wg)
+		}
+	}
+
+	go func() {
+		defer close(wrkchan)
+		wg.Wait()
+	}()
+
+	for n := range wrkchan {
+		if n != nil {
+			err = n
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("!! Done updating profile !!") // TODO remove
+
+	return nil
+}
+
+func UpdateProfileTableItem(i string, v string, userID string, svc *dynamodb.DynamoDB, ch chan error, wg *sync.WaitGroup) {
```

Comment:
We should update all attributes at the same time: https://stackoverflow.com/a/51072244
(Then we also don't need the goroutines and synchronization mechanism)

Reply:
This was the only safe way to wrote the update function, and also make it fast.

Reply:
The reference is written in JavaScript.

Reply:
I tried updating at once but this became a problem "UpdateExpression: aws.String("set " + i + " = :r"),". I needed a way to generate the update statement (which is more like the query language) and make it generic. In the end, I still needed to loop. This solution made it less complex.